src/Momento.Sdk/Internal/ScsTopicClient.cs

```diff
@@ -132,72 +132,125 @@ private async Task<TopicPublishResponse> SendPublish(string cacheName, string to
     }
 
-    private async ValueTask<TopicMessage?> GetNextRelevantMessageFromGrpcStreamAsync(AsyncServerStreamingCall<_SubscriptionItem> subscription,
-        CancellationToken cancellationToken, string cacheName, string topicName)
+    private class SubscriptionWrapper : IDisposable
     {
-        if (cancellationToken.IsCancellationRequested)
+        private readonly TopicGrpcManager _grpcManager;
+        private readonly string _cacheName;
+        private readonly string _topicName;
+        private readonly CacheExceptionMapper _exceptionMapper;
+        private readonly ILogger _logger;
+
+        private AsyncServerStreamingCall<_SubscriptionItem> _subscription;
+        private ulong? _lastSequenceNumber;
+
+        public SubscriptionWrapper(TopicGrpcManager grpcManager, string cacheName,
+            string topicName, CacheExceptionMapper exceptionMapper, ILogger logger)
         {
-            return null;
+            _grpcManager = grpcManager;
+            _cacheName = cacheName;
+            _topicName = topicName;
+            _exceptionMapper = exceptionMapper;
+            _logger = logger;
+
+            _subscription = Subscribe();
         }
 
-        try
+        private AsyncServerStreamingCall<_SubscriptionItem> Subscribe()
+        {
+            var request = new _SubscriptionRequest
+            {
+                CacheName = _cacheName,
+                Topic = _topicName
+            };
+            if (_lastSequenceNumber != null)
+            {
+                request.ResumeAtTopicSequenceNumber = _lastSequenceNumber.Value;
+            }
+
+            _logger.LogTraceExecutingTopicRequest(RequestTypeTopicSubscribe, _cacheName, _topicName);
+            return _grpcManager.Client.subscribe(request, new CallOptions());
+        }
+
+        public async ValueTask<TopicMessage?> GetNextRelevantMessageFromGrpcStreamAsync(
+            CancellationToken cancellationToken)
         {
-            while (await subscription.ResponseStream.MoveNext(cancellationToken))
+            while (!cancellationToken.IsCancellationRequested)
             {
-                var message = subscription.ResponseStream.Current;
+                try
+                {
+                    await _subscription.ResponseStream.MoveNext(cancellationToken);
+                }
+                catch (OperationCanceledException)
+                {
+                    break;
+                }
+                catch (Exception e)
+                {
+                    var sdkException = _exceptionMapper.Convert(e);
+                    if (sdkException.ErrorCode == MomentoErrorCode.CANCELLED_ERROR)
+                    {
+                        break;
+                    }
+                    //TODO: Are there other errors that we should break for?
+
+                    _logger.LogTraceTopicSubscriptionError(_cacheName, _topicName, sdkException);
+
+                    Dispose();
+                    _subscription = Subscribe();
```

Comment:
The go sdk checks the connection state before reconnecting. I don't do that here because grpc web doesn't have a connection state to check, but we could conditionally try it.

Reply:
Why do you instantly fire the reconnect and then later await 5 seconds before using it? Shouldn't those go the other way around?

Reply:
🤦 I must have mixed those up when I was doing some testing. I'll switch that around.